Source/Core/Runtime/SceneObjects/ProcessSceneObject.cs

```diff
@@ -93,18 +121,82 @@ protected void Awake()
             }
         }
 
-        /// <inheritdoc />
-        public void SetUniqueId(Guid guid)
+#if UNITY_EDITOR
+        private void OnValidate()
         {
-            uniqueId = guid.ToString();
+            // similar to on Serialize, but gets called on Copying a Component or Applying a Prefab
+            if (IsAssetOnDisk())
+            {
+                serializedGuid = null;
+                guid = System.Guid.Empty;
+            }
         }
 
-        private void Reset()
+        /// <summary>
+        /// Implement this method to receive a callback before Unity serializes your object.
+        /// </summary> 
+        /// <remarks>
+        /// We use this to prevent the GUID to be saved into a prefab on disk.
+        /// Be aware this is called more often than you would think (e.g.: about once per frame if the object is selected in the editor)
+        /// - https://discussions.unity.com/t/onbeforeserialize-is-getting-called-rapidly/115546, 
+        /// - https://blog.unity.com/engine-platform/serialization-in-unity </remarks>
+        public void OnBeforeSerialize()
         {
-            Init();
-        }
 
 #if UNITY_EDITOR
+            // This lets us detect if we are a prefab instance or a prefab asset.
+            // A prefab asset cannot contain a GUID since it would then be duplicated when instanced.
+            if (IsAssetOnDisk())
+            {
+                serializedGuid = null;
+                guid = System.Guid.Empty;
+                return;
+            }
+
+#endif
+            if (IsGuidAssigned() && !serializedGuid.Equals(guid))
+            {
+                serializedGuid.SetGuid(guid);
+#if UNITY_EDITOR
+                //FIXME SetDirty does not work guid string is not updated
+                EditorUtility.SetDirty(this);
```

Comment:
I'm missing something / doing it wrong? Also does not work further down. (Just realizing it might be coming from ProcessSceneObjectEditor)